### DISCUSSION TITLE
frontend.Variable to big.Int or []byte

### DISCUSSION THREAD

**Discussion by LautaroLasorsa** [Consensys/gnark#770] [Q&A]
Hi, I have the following structure:

```go
type Height struct {
	RevisionHeight frontend.Variable `gnark:",public"`
	RevisionNumber frontend.Variable `gnark:",public"`
}

func (height *Height) Hash(api frontend.API) frontend.Variable {
	mimc, err := mimc.NewMiMC(api)
	if err != nil {
		panic(err)
	}
	mimc.Write(height.RevisionHeight)
	mimc.Write(height.RevisionNumber)
	return mimc.Sum()
}

type Data struct {
	Height           Height            `gnark:",public"`
	TrustedHeight    Height            `gnark:",public"`
	CurrentTimestamp frontend.Variable `gnark:",public"`
}

func (data *Data) Hash(api frontend.API) frontend.Variable {
	mimc, err := mimc.NewMiMC(api)
	if err != nil {
		panic(err)
	}
	mimc.Write(data.Height.Hash(api))
	mimc.Write(data.TrustedHeight.Hash(api))
	mimc.Write(data.CurrentTimestamp)
	return mimc.Sum()
}
```

and I have to addapt the following code from an Eddsa sample:

```go
	var msg big.Int
	msg.Rand(randomness, snarkField)
	msgDataUnpadded := msg.Bytes()
	msgData := make([]byte, len(snarkField.Bytes()))
	copy(msgData[len(msgData)-len(msgDataUnpadded):], msgDataUnpadded)
```

I need to adapt it to use the Data.Hash() output as msg, but it doesn't work with a frontend.Variable msg.

How can I adapt it?

Thanks in advance.

**Comment by ivokub** [comment]
Hi. When writing circuits there is the concept of computing in-circuit and computing out-circuit.

In-circuit you have to use https://pkg.go.dev/github.com/consensys/gnark@v0.8.0/std/hash/mimc#NewMiMC, but when computing the witness, you have to use the out-circuit equivalent https://pkg.go.dev/github.com/consensys/gnark-crypto@v0.11.0/hash

Can you post the full example which can be executed (as a test file would be best). I'm not fully following what you are trying to achieve.
